solaris: compile errors in getent and sun_streams

furiel opened the pull request:
There is a little API difference between the linux and solaris version of getent functions.

In Linux, they return errno, and data in an output parameter.
In solaris, they return the data, and set errno.

This patch provides wrapper around the solaris versions, so that we can use them as in linux: errno is returned, and data is written to the output parameter.

There was a minor compile error in sun_streams, it is fixed as well.

Fixes: https://github.com/balabit/syslog-ng/issues/1667 and https://github.com/balabit/syslog-ng/issues/1692

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

Kokan commented on modules/getent/getent-group.c:
```diff
@@ -20,6 +20,8 @@
  * COPYING for details.
  */
 
+#include "compat.h"
```
These files are included into the `tfgetent.c`, which is not the best idea imho.
But if you don't plan to resolve this incleption I think this `compat.h` should be in the `tfgentent.c`. Because it would mix the two concept.
Of course the preferred fix would be to make these files properly independent compilation chunk.

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

presidento commented:
Please, move the _compat_ functionality to `lib` folder.
And perhaps we can use a feature macro (like `GNU_SOURCE` or `BSD_SOURCE`.

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

furiel commented:
For now I just moved to lib.

About the feature macros:
There are no feature macros for these:
getprotobynumber_r
getprotobyname_r
getservbyport_r
getservbyname_r

But for these there are compatible declarations with feature macros:
getgrnam_r
getgrgid_r
getpwnam_r
getpwuid_r

Checked on Oracle Solaris 11.3

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

furiel commented:
For the feature macros, I added this to tfgetent.c
```
+#if defined(sun) || defined(__sun)
+#define _POSIX_PTHREAD_SEMANTICS
+#endif
```

However, as feature macros are involved now, I did not feel comfortable to test the group part only with compile so I added an extra commit: please check: `getent: crash with empty group` that clears: https://github.com/balabit/syslog-ng/issues/1692. This way I could manually test the feature.

presidento commented on lib/compat/getent.h:
```diff
@@ -0,0 +1,47 @@
+#include <pwd.h>
+#include <netdb.h>
+
+int _getprotobynumber_r(int proto,
```
For me, the underscore prefix is reserved for protected/private members. I suggest using either a different prefix (`bb_` can be good for me), or shallow the original function call in Solaris with a `#define`, like the one in [`pio.h`](https://github.com/balabit/syslog-ng/blob/master/lib/compat/pio.h#L41) file.

furiel replied:
@presidento thanks for comment, fixed.

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions
